### PR TITLE
Remove virtual calls in UDF Arrow Kernels

### PR DESCRIPTION
**Pull request by fedor-miron**
### Changelog entry 

Allow to pass reader and builder types to UnaryKernelExec and BinaryKernelExec to avoid virtual calls and enable autovectorization.

### Changelog category 

* Performance improvement

### CODE REVIEW

File: ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h [hunk 1]
```diff
@@ -62,13 +62,18 @@ class TUdfKernelState : public arrow::compute::KernelState {
 
         return *ScalarBuilder_;
     }
+    
+    const IValueBuilder& GetValueBuilder() {
```
Comment (vitstn) [inline ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h:66]: и где оно используется? непонятна цель изменений с pgbuilder

**Review by vitstn — DISMISSED** [review]
valuebuilder

File: ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h [hunk 2]
```diff
@@ -322,33 +327,54 @@ inline void PrepareSimpleArrowUdf(IFunctionTypeInfoBuilder& builder, TType* sign
...
+        auto* readerImpl = dynamic_cast<TReader*>(&reader);
+        Y_ENSURE(readerImpl, TStringBuilder() << "Got " << typeid(reader).name() << " as BlockReader");
+
         const auto& arg = batch.values[0];
         if (arg.is_scalar()) {
             auto& builder = state.GetScalarBuilder();
-            auto item = reader.GetScalarItem(*arg.scalar());
-            TDerived::Process(item, [&](TBlockItem out) {
+            auto item = readerImpl->GetScalarItem(*arg.scalar());
+            Process(item, state, [&](auto out) {
```
Comment (vitstn) [inline ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h:356]: зачем в process весь state? это выглядит утечкой имплементации

File: ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h [hunk 3]
```diff
@@ -359,33 +385,49 @@ struct TUnaryKernelExec {
     }
 };
 
-template <typename TDerived>
+template <typename TDerived, typename TReader1 = IBlockReader, typename TReader2 = IBlockReader, typename TBuilder = IArrayBuilder>
 struct TBinaryKernelExec {
```
Comment (igormunkin) [inline ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h:389]: It's worth adding `static_assert`s similar to the ones in `TUnaryKernelExec`.

File: ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h (hunk 1 above)
Comment (igormunkin) [inline ydb/library/yql/public/udf/arrow/udf_arrow_helpers.h:66]: Discussed offline: let's move these changes to the separate PR.

**Review by igormunkin — CHANGES_REQUESTED** [review]
@fedor-miron, thanks for the patch! Besides the nits below, I ask you to provide a PR where it's used, if you have one.

**Review by igormunkin — DISMISSED** [review]
@fedor-miron, thanks for the fixes! LGTM.

**Review by igormunkin — APPROVED** [review]

**Pull request merged**